File: Assets/Scripts/Core/Effects/VoteImmunityEffect.cs
namespace Bunker.Core.Effects
{
    // Grants the caster immunity from being eliminated by this round's vote.
    // Self-targeted only.
    public class VoteImmunityEffect : ISpecialCardEffect
    {
        public bool CanApply(GameSession session, PlayerData caster, PlayerData target)
        {
            return session.Phase == GamePhase.Reveal || session.Phase == GamePhase.Discussion;
            // Must be used before voting starts — using it during/after Voting
            // would be unfair since results might already be known.
        }

        public void Apply(GameSession session, PlayerData caster, PlayerData target)
        {
            caster.HasVoteImmunityThisRound = true;
        }
    }
}

File: bunker-server/src/game/effects/VoteImmunityEffect.ts
import { ISpecialCardEffect } from "./ISpecialCardEffect";
import { GameSession } from "../GameSession";
import { PlayerData } from "../PlayerData";
import { GamePhase } from "../GamePhase";

// Grants the caster immunity from being eliminated by this round's vote.
// Self-targeted only, must be used before voting starts.
export class VoteImmunityEffect implements ISpecialCardEffect {
  canApply(session: GameSession, _caster: PlayerData, _target: PlayerData | null): boolean {
    return session.phase === GamePhase.Reveal || session.phase === GamePhase.Discussion;
  }

  apply(_session: GameSession, caster: PlayerData, _target: PlayerData | null): void {
    caster.hasVoteImmunityThisRound = true;
  }
}
